### PR TITLE
Add aria-describedby if image has no alt but has caption

### LINKED ISSUE #42057: Image block accessibility improvements
## What problem does this address?

Currently if the Image does not have an alt tag, but does have a caption there is no semantic link between the image and caption in order for screen readers, etc. to make use of it. 

Also, if an image is linked there is no semantic association between the link and image.

## What is your proposed solution?

Use an aria-describedby to point to the existing caption if there is a caption. If there is no labeling information at all, we should not insert any; we just need to ensure that if there is labeling information available, it should be associated with the image, or the link if the image is linked.

### PR DESCRIPTION
**Pull request by carolinan**
## What?

If an image block has no alt text but has a caption, use the caption to describe the image.
If the image is linked but has no alt text, use the caption to describe the link.

Fixes  https://github.com/WordPress/gutenberg/issues/42057
## Why?

To add context to the image and link when alt text is missing.

## How?

Uses the image id to create an id that is added the caption. This could be improved since it is not truly unique.
The id is then used with the aria-describedby that is placed on the href and the image.

**Alternatives:**
- Add the caption to the alt attribute when the alt is missing-
- Use aria-labeledby to create an accessible name. However, the described by is more suitable for longer texts, and we don't know how much content is in the caption.

## Testing Instructions

1. Add an image block that does not have an alt text.
2. Add a caption to it and duplicate the block.
3. Please test both linked and unlink images.

View the source to confirm that:
- An id is added to the figcaption.
- The image and the link has an aria-describedby attribute that uses the id.
Also view the front and confirm that the attributes are correct.

Please test the linked image with a screen reader and confirm that the caption is announced when you navigate to the link.

On one block, add an alt text and confirm that the ID and aria-describedby attributes are not present.
On the second block, (which does not have an alt text) remove the caption and confirm that the ID and aria-describedby attributes are not present.

### CODE REVIEW

**Comment by carolinan** [comment]
I am not able to update the snapshots, I keep getting this error:
```
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^
FetchError {
  message: 'request to http://localhost:8889/ failed, reason: connect ECONNREFUSED ::1:8889',
  type: 'system',
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED'
}
```
I am using wp-env.

**Comment by alexstine** [comment]
@carolinan Cannot say when, but I will try to jump in and help with this soon.

**Comment by alexstine** [comment]
@carolinan The code looks good to me. Can you tell me from logs which snapshots need to be updated and I can update those? Parsing some of the output from these failing tests can be tricky. 😞

**Comment by carolinan** [comment]
The native tests are failing.
Snapshot name: `Gallery block sets caption to gallery items 1`
file: `packages/block-library/src/gallery/test/__snapshots__/index.native.js.snap`

**Comment by alexstine** [comment]
@carolinan I refreshed the snapshot but tests still failing. Maybe further adjustments needed? If not, maybe test is flaky and just needs running a little more? This branch was super far behind trunk, that could have been part of the problem. It would not update for me at first either.

**Comment by carolinan** [comment]
@dcalhoun Hi, I am struggling with making the HTML attribute changes in this pull request pass the native E2E tests. I think I am only making it worse with each PR 😬 . Who is the best person to reach out to for help?

File: packages/block-library/src/image/save.js [hunk 1]
```diff
@@ -46,10 +46,17 @@ export default function save( { attributes } ) {
 		[ `wp-image-${ id }` ]: !! id,
 	} );
 
+	const describedById = `wp-image-caption-${ id }`;
```
Comment (dcalhoun) [inline packages/block-library/src/image/save.js]: From my testing, image `id` can be `undefined` when setting an Image block's media via "Insert from URL." Therefore, this reference can become `wp-image-caption-undefined`. Similar to guards elsewhere referenced below, should we guard against this case to only apply this value if there is an `id` set? 

https://github.com/WordPress/gutenberg/blob/30526b0e397f81b33ce349ef0086a98e8053ebb7/packages/block-library/src/image/save.js#L45-L47

If we add a guard here, we should likely apply it elsewhere for `describedById`, e.g. `packages/block-library/src/image/image.js`.

**Review by dcalhoun — COMMENTED** [review]
> @dcalhoun Hi, I am struggling with making the HTML attribute changes in this pull request pass the native E2E tests. I think I am only making it worse with each PR 😬 . Who is the best person to reach out to for help?

Hey @carolinan. 👋🏻 Thanks for taking the time to address the failures and reaching out. 

The CI logs showcase the e2e testing device timing out. From debugging this locally, it appears the sheer number of block validation errors and subsequent transforms caused the device to become unresponsive. Obviously, such device freeze should not happen regardless the number of validation errors, but that is an issue outside the scope of this PR. 😄 

To address the issue for this PR, we should update the `packages/react-native-editor/src/initial-html.js` file so that the fixture data matches the updated block structure. This will prevent block validation errors from occurring. 

<details><summary>Patch Updating the native Demo editor fixtures</summary>

```diff
diff --git a/packages/react-native-editor/src/initial-html.js b/packages/react-native-editor/src/initial-html.js
index ddea8594b7..151ad50461 100644
--- a/packages/react-native-editor/src/initial-html.js
+++ b/packages/react-native-editor/src/initial-html.js
@@ -59,8 +59,8 @@ else:
 <figure class="wp-block-image"><img alt=""/></figure>
 <!-- /wp:image -->
 
-<!-- wp:image -->
-<figure class="wp-block-image"><img src="https://cldup.com/cXyG__fTLN.jpg" alt=""/><figcaption class="wp-element-caption">Mountain</figcaption></figure>
+<!-- wp:image {"id":1,"sizeSlug":"large"} -->
+<figure class="wp-block-image size-large"><img class="wp-image-1" src="https://cldup.com/cXyG__fTLN.jpg" alt="" aria-describedby="wp-image-caption-1"/><figcaption class="wp-element-caption" id="wp-image-caption-1">Mountain</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:video {"id":683} -->
@@ -81,67 +81,67 @@ else:
 
 <!-- wp:gallery {"columns":8,"linkTo":"none","className":"alignfull"} -->
 <figure class="wp-block-gallery has-nested-images columns-8 is-cropped alignfull"><!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon.png" alt=""/><figcaption class="wp-element-caption">Paragraph</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Paragraph</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Heading.png" alt=""/><figcaption class="wp-element-caption">Heading</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Heading.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Heading</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Subheading.png" alt=""/><figcaption class="wp-element-caption">Subheading</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Subheading.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Subheading</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Quote.png" alt=""/><figcaption class="wp-element-caption">Quote</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Quote.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Quote</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Image.png" alt=""/><figcaption class="wp-element-caption">Image</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Image.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Image</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Gallery.png" alt=""/><figcaption class="wp-element-caption">Gallery</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Gallery.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Gallery</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Cover-Image.png" alt=""/><figcaption class="wp-element-caption">Cover Image</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Cover-Image.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Cover Image</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Video.png" alt=""/><figcaption class="wp-element-caption">Video</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Video.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Video</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Audio.png" alt=""/><figcaption class="wp-element-caption">Audio</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Audio.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Audio</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Column.png" alt=""/><figcaption class="wp-element-caption">Columns</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Column.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Columns</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-File.png" alt=""/><figcaption class="wp-element-caption">File</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-File.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">File</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Code.png" alt=""/><figcaption class="wp-element-caption">Code</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Code.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Code</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-List.png" alt=""/><figcaption class="wp-element-caption">List</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-List.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">List</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Button.png" alt=""/><figcaption class="wp-element-caption">Button</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Button.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Button</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Embeds.png" alt=""/><figcaption class="wp-element-caption">Embeds</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-Embeds.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">Embeds</figcaption></figure>
 <!-- /wp:image -->
 
 <!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
-<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-More.png" alt=""/><figcaption class="wp-element-caption">More</figcaption></figure>
+<figure class="wp-block-image size-large"><img src="https://wordpress.org/gutenberg/files/2018/07/Block-Icon-More.png" alt="" aria-describedby="wp-image-caption-undefined"/><figcaption class="wp-element-caption" id="wp-image-caption-undefined">More</figcaption></figure>
 <!-- /wp:image --></figure>
 <!-- /wp:gallery -->
 
```

</details>

Also, I left one comment regarding contexts where `id` would not be defined for Image blocks. If we modify the way we manage that context, we would need to update the native Demo editor fixtures to match the modification. 

Hope this helps!

**Comment by carolinan** [comment]
@dcalhoun I saw you also tried to re-run the test. Is the long hash for the id causing the timeout?

**Comment by dcalhoun** [comment]
@carolinan the emulator boot timeout appears to be occurring on `trunk` as well. So, I believe it is likely unrelated to the changes in this PR. It is rather unfortunate, but I am not sure it should block this work.

**Comment by aristath** [comment]
@WordPress/native-mobile Does anyone have any idea why the tests here are failing? 🤔

**Comment by geriux** [comment]
> @WordPress/native-mobile Does anyone have any idea why the tests here are failing? 🤔

I'm checking it out 👍

**Comment by geriux** [comment]
Hey there 👋 this should be fixed with this fix https://github.com/WordPress/gutenberg/pull/47643 which is already merged in `trunk` let us know if it solves the issue after bringing that change into this branch.

Thank you!

**Comment by carolinan** [comment]
Tests are now passing, but I don't know it if is correct that the pre existing core__image__center-caption fixtures were updated, or if that indicates that there is still a problem in deprecated.js.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for all the effort on this PR @carolinan, I know wrestling with deprecations isn't always a lot of fun! 🙇 

> I don't know it if is correct that the pre existing core__image__center-caption fixtures were updated

Given the changes proposed by this PR, the `core__image__center-caption` fixtures _would_ need updating. My understanding is that the fixture is covering that a normal, centered caption is parsed and serialized correctly without a deprecation involved. 

Having said that, we'd benefit from having a non-deprecation fixture that covered an image with `alt` text. It could be the original `core__image__center-caption` fixture modified to do this or a dedicated new fixture, e.g. `core__image__with-alt-text` etc.

While looking at the fixtures, I noticed an issue with the deprecation itself as well. It uses old, out-of-date `attributes` and `supports`. The deprecation needs to have the `attributes` and `supports` as they were immediately before the changes requiring the deprecation. 

In this case, the differences in the attribute definitions are only the inclusion of the `__experimentalRole` property for content-related attributes. I don't think this particularly matters for the deprecation. The out-of-date block supports config is more of a problem, however. The image block currently opts into more border supports than specified in the deprecation. I believe this may result in some border style attributes not being parsed and, therefore, not reapplied as the block is migrated to the latest version.

While having a deprecation to add the `aria-describedby` is great, users would still have to open their posts in the editor and re-save them for the image block to be migrated. 

Do you think it would be worthwhile in the image block's PHP render callback to check for a missing `aria-describedby` there and add it if necessary?

---

### TL;DR

- Yes, `core__image__center-caption` fixtures needed updating
- We should probably add a dedicated fixture that covers having `alt` text
- The deprecation's `attributes` and `supports` config objects are incorrect
- Users still need to edit posts with images and re-save them to get the `aria-describedby` added. Could this be injected at runtime if required?

I hope all that helped answer your question @carolinan 🙂

**Comment by carolinan** [comment]
Do I update the blockAttributes and blockSupports constants, or will that break the previous deprecations?

**Comment by aaronrobertshaw** [comment]
Each deprecation should have its attributes and supports as they were when the deprecated version was the latest version. 

The [handbook](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/) touches on this with the following:
>Deprecations are defined on a block type as its deprecated property, an array of deprecation objects where each object takes the form:
>
> - attributes (Object): The [attributes definition](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/) of the **deprecated form** of the block.
> - supports (Object): The [supports definition](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/) of the **deprecated form** of the block.

The wording could probably be clearer. If you have ideas on how best to clarify that, I'd be happy to review a PR to do so.

> Do I update the blockAttributes and blockSupports constants, or will that break the previous deprecations?

If you update those constants, then the older deprecations' attributes and supports wouldn't be as they were for those deprecated forms of the block. That might not always break the deprecations but it would only confuse things further.

I understand reducing the duplicate code in these deprecations, but those constants do lead to the assumption that they can always be reused, which isn't exactly true.

**Comment by carolinan** [comment]
> I understand reducing the duplicate code in these deprecations, but those constants do lead to the assumption that they can always be reused, which isn't exactly true.

Yes it is confusing.

File: packages/block-library/src/image/index.php [hunk 2]
```diff
@@ -11,9 +11,25 @@
...
  * @param  array  $attributes The block attributes.
  * @param  string $content    The block content.
- * @return string Returns the block content with the data-id attribute added.
+ * @return string Returns the block content with the data-id attribute and aria-describedby added.
  */
 function render_block_core_image( $attributes, $content ) {
+
+	// If an image block has no alternative text but has a caption,
+	// and aria-describedby is not set, add aria-describedby to the image or image link.
+	if ( empty( $attributes['alt'] ) &&
```
Comment (carolinan) [inline packages/block-library/src/image/index.php]: This is not working because `$attributes['alt']` is always empty, even if the user has added a text. I am testing other solutions.

**Comment by carolinan** [comment]
Deprecated.js
- [x] The supports are added as they were before the deprecation.
- [x] The attributes are added as they were before the deprecation.
- [x] The save function is added as it was before the deprecation.

Migration: Not used.

Fixtures
- [x] Updated: core__image__center-caption
- [x] New: core__image__deprecated-center-caption-6, core__image__with-alt-text

Snapshots
- [x] Updated and uses the new format without escaping

File: packages/block-library/src/image/deprecated.js [hunk 3]
```diff
@@ -93,6 +98,181 @@ const blockSupports = {
...
+				width: true,
+				__experimentalSelector: 'img, .wp-block-image__crop-area',
+				__experimentalSkipSerialization: true,
+				__experimentalDefaultControls: {
+					color: true,
+					radius: true,
+					width: true,
+				},
+			},
+		},
```
Comment (aaronrobertshaw) [inline packages/block-library/src/image/deprecated.js]: These supports don't match those currently in the Image block.json on trunk. 

The `__experimentalSelector` I don't think is critical but the duotone support was stabilized under `filter.duotone`. This should probably match that.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for continuing on this PR @carolinan 👍 

It looks like it is pretty close, just a couple of minor things to fix up.

- The supports definition in the new deprecation doesn't match the supports on trunk
- There's a conflict in the index.php after the recent Image updates to support the interactivity API and lightbox behaviour.

✅ Deprecation's attributes, and save function, match trunk
✅ Fixtures look good and fixture tests pass
✅ Old image blocks with caption but no alt migrate correctly in the editor
✅ Deprecated image blocks still render correctly on the frontend
✅ New image blocks behave as expected

**Comment by carolinan** [comment]
@artemiomorales Hi! I am going to need your help with reimplementing this accessibility bug fix, after the introduction of the lightbox.
Because I did not fully understand why the aria labels were added the way they are without translations or why the caption is not supported.
I can visually see the caption when the lightbox is opened or closed, but it is never announced by screen readers.

**Comment by artemiomorales** [comment]
Hey @carolinan! Thanks for reaching out. I started taking a look at this but have been diverted to resolve [outstanding bugs](https://github.com/WordPress/gutenberg/issues/51132) on the [lightbox](https://github.com/WordPress/gutenberg/pull/50373), and am hoping to get back to this as soon as I can.

I see that the translation issues [have been fixed](https://github.com/WordPress/gutenberg/pull/50962), and in my testing so far, I see that the captions are indeed announced by default using NVDA on Windows but not using VoiceOver on Mac. Which screen reader were you testing with?

In any case, do you still need my assistance on this? I'm thinking I'd have bandwidth for this likely next week.
